tree/ntuple/v7/inc/ROOT/RNTupleIndex.hxx

```diff
@@ -0,0 +1,141 @@
+/// \file ROOT/RNTupleIndex.hxx
+/// \ingroup NTuple ROOT7
+/// \author Florine de Geus <florine.de.geus@cern.ch>
+/// \date 2024-04-02
+/// \warning This is part of the ROOT 7 prototype! It will change without notice. It might trigger earthquakes. Feedback
+/// is welcome!
+
+/*************************************************************************
+ * Copyright (C) 1995-2024, Rene Brun and Fons Rademakers.               *
+ * All rights reserved.                                                  *
+ *                                                                       *
+ * For the licensing terms see $ROOTSYS/LICENSE.                         *
+ * For the list of contributors see $ROOTSYS/README/CREDITS.             *
+ *************************************************************************/
+
+#ifndef ROOT7_RNTupleIndex
+#define ROOT7_RNTupleIndex
+
+#include <ROOT/RField.hxx>
+#include <ROOT/RNTupleUtil.hxx>
+
+#include <unordered_map>
+#include <memory>
+#include <set>
+#include <string>
+#include <vector>
+
+namespace ROOT {
+namespace Experimental {
+namespace Internal {
+/////////////////////////////////////////////////////////////////////////////
+/// Container for the combined hash of the indexed fields. Uses the implementation from `boost::hash_combine` (see
+/// https://www.boost.org/doc/libs/1_55_0/doc/html/hash/reference.html#boost.hash_combine).
+struct RIndexValue {
+   std::size_t fValue = 0;
+
+   void operator+=(std::size_t other) { fValue ^= other + 0x9e3779b9 + (fValue << 6) + (fValue >> 2); }
+   inline bool operator==(const RIndexValue &other) const { return other.fValue == fValue; }
+   inline size_t operator()(const ROOT::Experimental::Internal::RIndexValue &val) const { return val.fValue; }
+};
+
+// clang-format off
+/**
+\class ROOT::Experimental::Internal::RNTupleIndex
+\ingroup NTuple
+\brief Build an index for an RNTuple so it can be joined onto other RNTuples.
+*/
+// clang-format on
+class RNTupleIndex {
+   friend std::unique_ptr<RNTupleIndex>
+   CreateRNTupleIndex(std::vector<std::string_view> fieldNames, RPageSource &pageSource);
```

Comment:
This `std::vector` should probably be either a const ref or a `std::span` to avoid deep-copying it.